database.cc

```diff
@@ -2268,6 +2268,69 @@ future<> database::clear_snapshot(sstring tag, std::vector<sstring> keyspace_nam
     });
 }
 
+future<std::unordered_map<sstring, std::unordered_map<sstring, db::snapshot_ctl::snapshot_details>>> database::get_snapshot_details() {
+    return seastar::async([this] {
+        std::vector<sstring> data_dirs = _cfg.data_file_directories();
+        auto dirs_only_entries_ptr =
+            make_lw_shared<lister::dir_entry_types>(lister::dir_entry_types{directory_entry_type::directory});
+        std::unordered_map<sstring, std::unordered_map<sstring, snapshot_ctl::snapshot_details>> details;
+
+        for (auto& datadir : data_dirs) {
+            lister::scan_dir(datadir, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details] (fs::path parent_dir, directory_entry de) mutable {
+                // KS directory
+                sstring ks_name = de.name;
+
+                return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) mutable {
+                    // CF directory
+                    sstring cf_name = de.name.substr(0, de.name.find("-"));
+                    fs::path cf_dir = parent_dir / de.name;
+
+                    return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, dirs_only_entries_ptr, &details, ks_name, cf_name = std::move(cf_name), cf_dir = std::move(cf_dir)] (fs::path parent_dir, directory_entry de) mutable {
+                        // "snapshots" directory
+
+                        return lister::scan_dir(parent_dir / de.name, *dirs_only_entries_ptr, [this, &details, ks_name, cf_name, cf_dir] (fs::path parent_dir, directory_entry de) {
+                            sstring snapshot_name = de.name;
+                            sstring cf_map_key = ks_name + ":" + cf_name;
```

Comment:
Can we have a pair of strings instead of combining them with a colon?
While the colon character is technically not allowed in table names in CQL, it's not outlawed by our underlying code, and Alternator does uses the colon character in some of the table names (for materialized views).

Reply:
yeah, sure